header.go

```diff
@@ -0,0 +1,84 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package yarpc
+
+import "github.com/yarpc/yarpc-go/internal"
+
+// CanonicalizeHeaderKey canonicalizes the given header key for storage into
+// the Headers map.
+func CanonicalizeHeaderKey(k string) string {
+	return internal.CanonicalizeHeaderKey(k)
+}
+
+// Headers defines application headers which will be sent over the wire to the
+// recipient of an RPC call.
+type Headers internal.Headers
+
+// NewHeaders builds a new Headers object.
+func NewHeaders() Headers {
+	return Headers{}
+}
+
+// With returns a Headers object with the given key-value pair added to it.
+// If a header with the same name already exists, it will be overwritten.
+//
+// The returned Headers object MAY use a different underlying data store than
+// the original Headers object, and so the returned object MUST always be used
+// instead of the original object.
+//
+// 	headers = headers.With("foo", "bar")
+//
+// This call may be chained to set multiple headers consecutively.
+//
+// 	headers = headers.With("foo", "bar").With("baz", "qux")
+func (h Headers) With(k, v string) Headers {
+	return Headers(internal.Headers(h).With(k, v))
+}
+
+// Get retrieves the value associated with the given header key, and a
+// boolean indicating whether the key actually existed in the header map.
+func (h Headers) Get(k string) (string, bool) {
+	return internal.Headers(h).Get(k)
+}
+
+// Del deletes the given header key from the map.
+func (h Headers) Del(k string) {
+	internal.Headers(h).Del(k)
+}
+
+// Len returns the number of headers defined on this object.
+func (h Headers) Len() int {
+	return internal.Headers(h).Len()
+}
+
+// Keys returns a list of header keys defined on this Headers object.
```

Comment:
should this method sort the keys so that it's deterministic  and consistent? Otherwise, call out that the keys are in an undefined order (which is a little annoying to use)